File: CSC310-Cpp/_solutions/Module07-Greedy-Algorithms/debug01/activity_selection.cpp
// activity_selection.cpp
// Greedy activity selection (CLRS 15.1). Each activity has a start time and a
// finish time; two activities are compatible if their half-open intervals
// [start, finish) do not overlap. The goal is a maximum-size set of mutually
// compatible activities.
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

struct Activity {
    std::string name;
    int start;
    int finish;
};

// Returns a maximum set of compatible activities, in the order chosen.
std::vector<Activity> selectActivities(std::vector<Activity> acts) {
    std::sort(acts.begin(), acts.end(), [](const Activity& a, const Activity& b) {
        return a.finish < b.finish;
    });

    std::vector<Activity> chosen;
    if (acts.empty()) {
        return chosen;
    }
    chosen.push_back(acts[0]);
    int lastFinish = acts[0].finish;
    for (std::size_t i = 1; i < acts.size(); ++i) {
        if (acts[i].start >= lastFinish) {
            chosen.push_back(acts[i]);
            lastFinish = acts[i].finish;
        }
    }
    return chosen;
}

void report(const std::string& label, const std::vector<Activity>& acts) {
    std::vector<Activity> chosen = selectActivities(acts);
    std::cout << label << ": " << chosen.size() << " activities:";
    for (const Activity& a : chosen) {
        std::cout << ' ' << a.name << "[" << a.start << "," << a.finish << ")";
    }
    std::cout << '\n';
}

int main() {
    std::vector<Activity> clrs = {
        {"a1", 1, 4},  {"a2", 3, 5},  {"a3", 0, 6},   {"a4", 5, 7},
        {"a5", 3, 9},  {"a6", 5, 9},  {"a7", 6, 10},  {"a8", 8, 11},
        {"a9", 8, 12}, {"a10", 2, 14}, {"a11", 12, 16}};
    report("textbook set", clrs);

    std::vector<Activity> room = {
        {"lecture", 9, 10}, {"seminar", 8, 13}, {"lab", 10, 11},
        {"office-hours", 11, 12}, {"club", 12, 13}};
    report("room schedule", room);

    std::vector<Activity> nested = {
        {"all-day", 0, 10}, {"short-1", 1, 2}, {"short-2", 2, 3}, {"short-3", 3, 4}};
    report("nested set", nested);
    return 0;
}

File: CSC310-Python/Module07-Greedy-Algorithms/debug01/activity_selection.py
"""Activity selection (CLRS Section 15.1).

Given activities with start and finish times, choose a maximum-size set
of mutually compatible activities (no two overlap in time).  The greedy
procedure sorts the activities and then repeatedly takes the first one
that is compatible with the last one chosen.
"""


def select_activities(activities):
    """activities: list of (name, start, finish).  Returns chosen names."""
    ordered = sorted(activities, key=lambda a: a[1])
    chosen = [ordered[0]]
    last_finish = ordered[0][2]
    for act in ordered[1:]:
        name, start, finish = act
        if start >= last_finish:
            chosen.append(act)
            last_finish = finish
    return [a[0] for a in chosen]


def brute_force_best(activities):
    """Largest compatible subset by trying all subsets (for checking only)."""
    n = len(activities)
    best = 0
    for mask in range(1 << n):
        subset = [activities[i] for i in range(n) if mask >> i & 1]
        subset.sort(key=lambda a: a[1])
        ok = all(subset[i][2] <= subset[i + 1][1] for i in range(len(subset) - 1))
        if ok:
            best = max(best, len(subset))
    return best


def report(title, activities):
    chosen = select_activities(activities)
    print(f"{title}")
    print(f"  chosen  : {chosen}")
    print(f"  count   : {len(chosen)}   (optimum {brute_force_best(activities)})")


def main():
    # The eleven activities of the CLRS Section 15.1 example
    starts = [1, 3, 0, 5, 3, 5, 6, 8, 8, 2, 12]
    finishes = [4, 5, 6, 7, 9, 9, 10, 11, 12, 14, 16]
    clrs = [(f"a{i + 1}", s, f) for i, (s, f) in enumerate(zip(starts, finishes))]
    report("CLRS activity set", clrs)

    lab = [("setup", 0, 3), ("intro", 1, 2), ("lecture", 2, 5), ("demo", 4, 6),
           ("quiz", 5, 7), ("review", 6, 9), ("wrapup", 8, 10), ("long", 0, 9)]
    report("Lab schedule", lab)


if __name__ == "__main__":
    main()
